README.md

```diff
@@ -1 +1,292 @@
+
+## Exporting Functions
+
+To make the module smarter (more dynamic), we need to use the `module.exports = function() {}` pattern which is more versatile. It allows us to execute some logic and even use parameters from the function's arguments.
+
+With the function pattern, we can re-write our configuration module to do more things. In other words, the module will can behave differently because we can put some logic in its code. This is `environment.js` file. The first thing in it is getting the environment or using the one from the environment variable `NODE_ENV` if the argument is not set:
+
+```js
+// environment.js
+module.exports = function(env) {
+  env = env || process.env.NODE_ENV
+```
+
+Then we define two conditions and return different configurations accordingly (in this example only API keys are different):
+
+```js
+  if (env == 'production') {
+    return {
+      url: 'http://webapplog.com',
+      name: 'React Quicly',
+      port: 3000,
+      apiKey: "34EC5CE9-CD02-4C4E-B5D9-D7CEF2F12F2C"
+    }
+  } else if (env == 'testing') {
+    return {
+      url: 'http://webapplog.com',
+      name: 'React Quicly',
+      port: 3000,
+      apiKey: "D2230FC4-7E2A-4003-9372-0BF80107558D"
+    }
+  } else { // Assume development
+    return {
+      url: 'http://webapplog.com',
+      name: 'React Quicly',
+      port: 3000,
+      apiKey: "D1FCFE58-575F-40FF-AF36-0CBD6A367756"
+    }
+  }
+}
+```
+
+The `main-environment.js` file can pass the argument or not, but it needs to invoke the imported configs method (from `environment`):
+
+```js
+// main-environment.js
+var configsInit = require('./environment')
+configs = configsInit('your-awesome-environment')
+console.log('URL is ' + configs.url)
+console.log('API key is ' +  configs.apiKey)
+```
+
+or you can merge 1st and 2nd lines into a single statement (more common style):
+
+```js
+var configs = require('./environment')()
+console.log('URL is ' +  configs.url)
+console.log('API key is ' +  configs.apiKey)
+```
+
+The resulting module and main file will produce be able to switch between different settings according to the environment variable:
+
+```
+$ node main-environment
+URL is http://webapplog.com
+API key is D1FCFE58-575F-40FF-AF36-0CBD6A367756
+$ NODE_ENV=production node main-environment
+URL is http://webapplog.com
+API key is 34EC5CE9-CD02-4C4E-B5D9-D7CEF2F12F2C
+$ NODE_ENV=testing node main-environment
+URL is http://webapplog.com
+API key is D2230FC4-7E2A-4003-9372-0BF80107558D
+```
+
+To summarize the difference between objects and functions, the latter are more dynamic and allow for the so called initializer or constructor code, i.e., some code to tweak an object before exporting it.
+
+## module.exports and exports
+
+Simply put, `module.exports` is an alias to the `exports`. You can save your self a repetitive stress injury by using `exports`. Take a look at this module which uses `module.exports`:
```

Comment:
Well, not exactly. "Alias" implies that they are always equivalent, but it's important for students to understand that the `module` object and the `export` object are independent -- thus, changing `module.exports` to something other than `exports` is what causes the issues that are explored in this README. That's why I added "to start".